chain-evm/src/machine.rs

```diff
@@ -98,6 +112,73 @@ fn precompiles(fork: HardFork) -> Precompiles {
     }
 }
 
+impl<'runtime> VirtualMachine<'runtime> {
+    fn execute_transaction<F, T>(
+        &mut self,
+        gas_limit: u64,
+        delete_empty: bool,
+        f: F,
+    ) -> Result<(&AccountTrie, &LogsState, T), Error>
+    where
+        F: Fn(
+            &mut StackExecutor<
+                'runtime,
+                '_,
+                MemoryStackState<'_, 'runtime, VirtualMachine>,
+                Precompiles,
+            >,
+            u64,
+        ) -> (ExitReason, T),
+    {
+        let executable = |gas_limit| {
+            let metadata = StackSubstateMetadata::new(gas_limit, self.config);
+            let memory_stack_state = MemoryStackState::new(metadata, self);
+            let mut executor = StackExecutor::new_with_precompiles(
+                memory_stack_state,
+                self.config,
+                &self.precompiles,
+            );
+            (f(&mut executor, gas_limit), executor)
+        };
+
+        let ((exit_reason, val), executor) = executable(gas_limit);
+        match exit_reason {
+            ExitReason::Succeed(_) => {
+                // apply and return state
+                // apply changes to the state, this consumes the executor
+                let state = executor.into_state();
+                // Next, we consume the stack state and extract the values and logs
+                // used to modify the accounts trie in the VirtualMachine.
+                let (values, logs) = state.deconstruct();
+
+                self.apply(values, logs, delete_empty);
+                // exit_reason
+                Ok((&self.state, &self.logs, val))
+            }
+
+            // WARNING: such approach was followed from the https://github.com/paritytech/frontier/blob/4a77d59d4640e492ae0db43f891929b75876cf46/client/rpc/src/eth.rs#L1447
+            // need to clarify it
```

Comment:
What exactly is needed to be clarified? From the docs I can see that we cannot really get an exact `ExitReason` that corresponds to `NotEnoughGas` which is, well, rather annoying. I'd rather have them make it into a separate error. I'll make it a separate task. Executing the contract twice opens up the potential for DDoS attacks.

Reply:
So your suggestion just proceed with that error info that we got from the EVM, without any clarification and later just update EVM side ?